substrate/frame/migrations/src/migrations.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Generic multi block migrations not specific to any pallet.
+
+use crate::{weights::WeightInfo, Config};
+use codec::Encode;
+use core::marker::PhantomData;
+use frame_support::{
+	migrations::{SteppedMigration, SteppedMigrationError},
+	traits::{OnGenesis, PalletInfoAccess},
+	weights::WeightMeter,
+};
+use sp_core::{twox_128, Get};
+use sp_io::{storage::clear_prefix, KillStorageResult};
+use sp_runtime::SaturatedConversion;
+
+/// Remove all of a pallet's state and re-initializes it to the current in-code storage version.
+///
+/// It uses the multi block migration frame. Hence it is safe to use even on
+/// pallets that contain a lot of storage.
+///
+/// # Parameters
+///
+/// - T: The runtime. Used to access the weight definition.
+/// - P: The pallet to resetted as defined in construct runtime
+///
+/// # Note
+///
+/// If your pallet does rely of some state in genesis you need to take care of that
+/// separately. This migration only sets the storage version after wiping by running
+/// [`OnGenesis::on_genesis`].
+pub struct ResetPallet<T, P>(PhantomData<(T, P)>);
+
+impl<T, P> ResetPallet<T, P>
+where
+	P: PalletInfoAccess,
+{
+	fn hashed_prefix() -> [u8; 16] {
+		twox_128(P::name().as_bytes())
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn num_keys() -> u64 {
+		let prefix = Self::hashed_prefix().to_vec();
+		crate::storage::KeyPrefixIterator::new(prefix.clone(), prefix, |_| Ok(())).count() as _
+	}
+}
+
+impl<T, P> SteppedMigration for ResetPallet<T, P>
+where
+	T: Config,
+	P: PalletInfoAccess + OnGenesis,
+{
+	type Cursor = bool;
+	type Identifier = [u8; 16];
+
+	fn id() -> Self::Identifier {
+		("RemovePallet::", P::name()).using_encoded(twox_128)
+	}
+
+	fn step(
+		cursor: Option<Self::Cursor>,
+		meter: &mut WeightMeter,
+	) -> Result<Option<Self::Cursor>, SteppedMigrationError> {
+		// we write the storage version in a seperate block
+		if cursor.unwrap_or(false) {
+			let required = T::DbWeight::get().writes(1);
+			if meter.remaining().any_lt(required) {
+				return Err(SteppedMigrationError::InsufficientWeight { required })
+			}
+			P::on_genesis();
```

Comment:
`on_genesis` has some narrow meaning, I would not use it here.

We should use `GetStorageVersion` and implement & seal a common trait for `NoVersionSet` and `StorageVersion` that gives you the function `write_to_storage` or something like that. While `NoVersionSet` would write `0` as `StorageVersion`.

Reply:
I agree that we probably should not use `OnGenesis` here. But isn't the point of this `NoVersionSet` that no version should be set in storage? Meaning when `NoVersionSet` is the associated type we simply should not write any version?

Reply:
> But isn't the point of this `NoVersionSet` that no version should be set in storage?

No. `0` is always the default. https://docs.rs/frame-support/latest/frame_support/traits/struct.NoStorageVersionSet.html the docs are not ultra clear :see_no_evil:  But basically this is just a hint for a user that tries to implement a migration and forgets to set a `storage_version`.

Reply:
Aye. Added the trait and using this now instead of `OnGenesis`. But I didn't seal as the related traits are also not sealed. Like `GetStorageVersion`. Do you think it is important to seal this one?

Reply:
@ggwpez Pointed out that we already have a trait for that. Switched to that one and removed my new trait.